## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [3]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [5]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [11]:
# Use the describe function to gather some basic statistics
df.describe()


<bound method Series.unique of 0              decor
1        consumables
2           software
3        consumables
4        consumables
            ...     
54634       software
54635    consumables
54636    consumables
54637      furniture
54638       software
Name: category, Length: 54639, dtype: object>

In [9]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# Select the first row of the DataFrame
df.iloc[0]


first                                    Donald
last                                    Harding
job                                Immunologist
phone                        793-904-7725x39308
email          harding.donald.7185@sullivan.com
client_id                                 58515
order_id                                8953482
order_date                           2023-04-28
order_week                                   17
order_year                                 2023
item_id                          EUD29711-63-6U
category                                  decor
subcategory                            wall art
unit_price                               1096.8
unit_cost                                762.71
unit_weight                                 7.5
qty                                         105
line_number                                   1
Name: 0, dtype: object

In [16]:
# Use this space to do any additional research
# What are the unique category names
df["category"].unique()

array(['decor', 'consumables', 'software', 'furniture', 'electronics'],
      dtype=object)

In [17]:
# Use this space to do any additional research
# What are the unique subcategory names
df["subcategory"].unique()

array(['wall art', 'pens', 'project management', 'misc', 'plants',
       'kitchen supplies', 'paper products', 'seating', 'computers',
       'stands', 'tables', 'bathroom supplies', 'audio visual',
       'accounting', 'logistics', 'telecom', 'desks', 'cabling'],
      dtype=object)

In [18]:
# Use this space to do any additional research
# Identify incomplete rows (Comment Post-Run:  There aren't any.)
df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [19]:
# Use this space to do any additional research
# Review datatypes by column (Post-Run:  OK)
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [21]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [65]:
# For the category with the most entries,
# which subcategory had the most entries?
# Create a new df with the category of interest, then count subcategories
cat_df = df.loc[df['category'] == 'consumables']
df['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [33]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
42481    206
71958    205
32500    202
49687    201
43607    197
Name: count, dtype: int64

In [51]:
# Store the client ids of those top 5 clients in a list.
# Source:  https://stackoverflow.com/questions/52441836/how-to-make-a-list-of-values-based-on-value-count-in-pandas
top_five_clients = df['client_id'].value_counts().index[:5].tolist()
print(top_five_clients)



[33615, 66037, 46820, 38378, 24741]


In [57]:
# How many total units (the qty column) did the
# client with the most entries order order?
# Part 1 :: Set the client number to a variable
top_client_number = df['client_id'].value_counts().index[:1].tolist()
top_client_number = top_client_number[0]
print(top_client_number)

33615


In [69]:
# Part 2 :: Set a new dataframe with topclient only
topclient_df = df.loc[(df["client_id"] == top_client_number)]
topclient_df.head()

first   last                                           job  \
5     Jessica  Reyes  Production designer, theatre/television/film   
12    Jessica  Reyes  Production designer, theatre/television/film   
724   Jessica  Reyes  Production designer, theatre/television/film   
840   Jessica  Reyes  Production designer, theatre/television/film   
1405  Jessica  Reyes  Production designer, theatre/television/film   

           phone                     email  client_id  order_id  order_date  \
5     5016915923  jessica.reyes@grimes.net      33615   1327730  2023-05-02   
12    5016915923  jessica.reyes@grimes.net      33615   6059361  2023-05-16   
724   5016915923  jessica.reyes@grimes.net      33615   7417275  2023-05-24   
840   5016915923  jessica.reyes@grimes.net      33615   1951089  2023-02-15   
1405  5016915923  jessica.reyes@grimes.net      33615   5615970  2023-04-19   

      order_week  order_year         item_id     category     subcategory  \
5             18        2023  BMB23158-88-9A        decor          plants   
12            20        2023  BDC65078-60-7M  consumables  paper products   
724           21        2023  ZBZ76765-83-2X  electronics         telecom   
840            7        2023  XBU40041-09-5M  consumables  paper products   
1405          16        2023  UCZ01380-66-1U  consumables  paper products   

      unit_price  unit_cost  unit_weight  qty  line_number  
5         172.84     154.44         2.79   78            0  
12        229.74     198.38        27.51  568            2  
724       222.03     155.57         1.33   26            2  
840         9.09       7.18         3.11   36            3  
1405       34.82      25.80         2.26   17            4

In [72]:
# Part 3 :: Sum "qty" for lead client
topclient_df["qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the 888888888888 HERE 8888888888
# subtotal for each line using the unit_price
# and the qty



unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.



unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [12]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%



line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [13]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).



first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [14]:
# Create a column for the profit of
# each line using line cost and line price



first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above


Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



client_id     qty  shipping_price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [19]:
# Sort the updated data by "Total Profit" form highest to lowest


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M